src/otx/core/model/entity/instance_segmentation.py

```diff
@@ -0,0 +1,149 @@
+# Copyright (C) 2023 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+#
+"""Class definition for instance segmentation model entity used in OTX."""
+
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, Any
+
+import numpy as np
+import torch
+from mmdet.registry import MODELS
+from mmdet.structures import DetDataSample
+from mmdet.structures.mask import BitmapMasks, PolygonMasks
+from mmengine.registry import MODELS as MMENGINE_MODELS
+from mmengine.structures import InstanceData
+from torchvision import tv_tensors
+
+from otx.core.data.entity.base import OTXBatchLossEntity
+from otx.core.data.entity.instance_segmentation import InstanceSegBatchDataEntity, InstanceSegBatchPredEntity
+from otx.core.model.entity.base import OTXModel
+from otx.core.utils.build import build_mm_model
+
+if TYPE_CHECKING:
+    from mmdet.models.data_preprocessors import DetDataPreprocessor
+    from omegaconf import DictConfig
+    from torch import nn
+
+
+class OTXInstanceSegModel(OTXModel[InstanceSegBatchDataEntity, InstanceSegBatchPredEntity]):
+    """Base class for the detection models used in OTX."""
+
+
+class MMDetInstanceSegCompatibleModel(OTXInstanceSegModel):
+    """Instance Segmentation model compatible for MMDet."""
+
+    def __init__(self, config: DictConfig) -> None:
+        self.config = config
+        self.load_from = self.config.pop("load_from", None)
+        super().__init__()
+
+    def _create_model(self) -> nn.Module:
+        det = MODELS.get("DetDataPreprocessor")
+        MMENGINE_MODELS.register_module(module=det)
+
+        return build_mm_model(self.config, MODELS, self.load_from)
+
+    def _customize_inputs(self, entity: InstanceSegBatchDataEntity) -> dict[str, Any]:
+        mmdet_inputs: dict[str, Any] = {}
+
+        mmdet_inputs["inputs"] = entity.images  # B x C x H x W PyTorch tensor
+        mmdet_inputs["data_samples"] = []
+
+        for img_info, bboxes, masks, polygons, labels in zip(
+            entity.imgs_info,
+            entity.bboxes,
+            entity.masks,
+            entity.polygons,
+            entity.labels,
+        ):
+            height, width = img_info.ori_shape
+            if len(masks):
+                mmdet_masks = BitmapMasks(
+                    masks.data.cpu().numpy(), height, width)
```

Comment:
Why do you send `masks` back from the GPU memory to the host memory? It will make this code blocking and bring undesired latency.

Reply:
I'm aware of this issue. The current structure of MMDet masks, as outlined in their implementation :https://github.com/open-mmlab/mmdetection/blob/main/mmdet/structures/mask/structures.py#L643 , doesn't accommodate `torch.tensor`. The structure is designed to handle only numpy arrays. Consequently, I'm required to detach and convert torch.tensor to a numpy array for compatibility.

We have two potential courses of action: invest time in implementing a customized BitmapMasks that supports torch.tensor, or continue with the existing approach of detaching to CPU and converting to numpy.

It's crucial to note that this issue extends to polygons as well. If we opt to switch from `list[datumaro.Polygon]` to `list[torch.FloatTensor]` in `InstanceSegDataEntity`, we will encounter the same challenge of detaching from GPU to CPU and converting to numpy.